srtcore/core.cpp

```diff
@@ -579,13 +575,11 @@ void CUDT::getOpt(SRT_SOCKOPT optName, void *optval, int &optlen)
         break;
 
     case SRTO_SENDER:
-        *(int32_t *)optval = m_config.bDataSender;
```

Comment:
1. As the doc suggested that it's safe for the user to pass a pointer to int and we cast it to bool *, except for the endian issue which also mentioned by the doc "just don't compare the result with 1".
2. It's unsafe to change the type from bool to int in the doc. The doc has been out there as the source code, it's not uncommon to only read the doc and didn't check the source code. It's hard to figure out the change in the doc if user has finished their work.

Make the code following the doc is safe in my opinion. It only sacrifice the flexibility to pass int * and "compare the result with 1".

Reply:
1. The actual byte size of `bool` is implementation-dependant and can be bigger than `int32_t` for all I know.
2. I bet the documentation has been out there for much less time than this particular piece of code in which case I would argue that it is the doc that has a bug :)

But technicalities aside, I think my main point is that for a minor release we probably should not change existing API behavior no matter how imperfect it seems to be unless it is an obvious crash and burn sort of situation.

Reply:
### 0. Options Changed
There are four socket options that are documented as using `bool`, but converting to `int` instead. See the PR description.

Were using `int` cast instead of `bool`:
- `SRTO_SENDER`
- `SRTO_TSBPDMODE`
- `SRTO_DRIFTTRACER`
- `SRTO_ENFORCEDENCRYPTION`

### 1. Boolean Type in GStreamer

GStreamer plugging seems to use two types for boolean options.

For example, `SRTO_TSBPDMODE` uses `G_TYPE_BOOLEAN` in the list of options, which is likely `int` ([see here](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/ext/srt/gstsrtobject.c#L168)), but `bool` type (1 byte?) for the default value (see [here](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/ext/srt/gstsrtobject.c#L150)).

- `srt_setsockopt` for `SRTO_TSBPDMODE` is called with `bool` size (see [here](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/ext/srt/gstsrtobject.c#L322)).

- `srt_setsockopt` for `SRTO_SENDER` is called with `bool` type (see [here](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/ext/srt/gstsrtobject.c#L1208)).

Luckily, `srt_getsockopt` is used only for `SRTO_PAYLOADSIZE` socket option (see [here](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/ext/srt/gstsrtobject.c#L1698)).

### 2. Change to bool or leave as is?

Casting to `bool` instead of `int` is safer, as at least it does not corrupt the application memory.
But has its own consequences:
- endianness, as mentioned by @quink-black;
- if `optval` was not set to `0` before calling `srt_getsockopt`, then three bytes of it may have some garbage, leading to `true` value (`!= 0`).

I am still not sure what's the best way to go, but corrupting memory looks worse than endianness and garbage when an application does not check the `optlen` returned.